Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyTreeView.cxx

```diff
@@ -810,44 +810,58 @@ void qMRMLSubjectHierarchyTreeView::setCurrentItems(QList<vtkIdType> items)
     return;
     }
 
-  this->clearSelection();
+  QSet<QModelIndex> itemsToSelect;
+  foreach (vtkIdType itemID, items)
+    {
+    QModelIndex itemIndex = d->SortFilterModel->indexFromSubjectHierarchyItem(itemID);
+    if (itemIndex.isValid())
+      {
+      itemsToSelect.insert(itemIndex);
+      }
+    }
 
-  foreach (long itemID, items)
+  const QModelIndexList selectedItemsList = this->selectionModel()->selectedRows();
+  QSet<QModelIndex> selectedItems(selectedItemsList.begin(), selectedItemsList.end());
+  QSet<QModelIndex> itemAlreadySelected = selectedItems;
+  itemAlreadySelected.intersect(itemsToSelect);
+
+  for (const QModelIndex &item : itemAlreadySelected)
+    {
+    selectedItems.remove(item);
```

Comment:
Overall, this looks too complex, it is just hard to read. Mostly because names are not accurate and/or the content changes but the name remains the same. There are also some unnecessary loops. Proposed implementation (not tested, just writing the code blindly):

```
  // Get requested selection
  QSet<QModelIndex> requestedSelectedItems;
  foreach (vtkIdType itemID, items)
    {
    QModelIndex itemIndex = d->SortFilterModel->indexFromSubjectHierarchyItem(itemID);
    if (itemIndex.isValid())
      {
      requestedSelectedItems.insert(itemIndex);
      }
    }

  // Get previous selection
  const QModelIndexList previouslySelectedItemsList= this->selectionModel()->selectedRows();
  QSet<QModelIndex> previouslySelectedItems(previouslySelectedItemsList.begin(), previouslySelectedItemsList.end());

  // Deselect items that were previously selected but not requested anymore
  foreach (QModelIndex itemIndex, previouslySelectedItems)
    {
    if (itemIndex.isValid() && !requestedSelectedItems.contains(itemIndex))
      {
      this->selectionModel()->select(itemIndex, QItemSelectionModel::Deselect | QItemSelectionModel::Rows);
      }
    }

  // Select items that are requested but have not been previously selected
  foreach (QModelIndex itemIndex, requestedSelectedItems)
    {
    if (!previouslySelectedItems.contains(itemIndex))
      {
      this->selectionModel()->select(itemIndex, QItemSelectionModel::Select | QItemSelectionModel::Rows);
      }
    }
```

Reply:
Thank you for your review, I've made the suggested improvements.